core/drivers/firewall/stm32_rif.c

```diff
@@ -0,0 +1,141 @@
+// SPDX-License-Identifier: BSD-2-Clause
+/*
+ * Copyright (c) 2021-2024, STMicroelectronics
+ */
+
+ #include <drivers/stm32_rif.h>
+ #include <kernel/panic.h>
+
+TEE_Result stm32_rif_check_access(uint32_t cidcfgr,
+				  uint32_t semcr,
+				  unsigned int nb_cid_supp,
+				  unsigned int cid_to_check)
+{
+	uint32_t scid_mask = 0;
+	uint32_t msb_nb_cid_supp = sizeof(nb_cid_supp) * 8 -
+				   __builtin_clz(nb_cid_supp | 1) - 1;
+
+	/* SCID bitfield highend can't be > SCID_SHIFT + MAX_CID_BITFIELD */
+	assert(msb_nb_cid_supp < MAX_CID_BITFIELD);
+
+	scid_mask = GENMASK_32(SCID_SHIFT + msb_nb_cid_supp, SCID_SHIFT);
+
+	if (!(cidcfgr & _CIDCFGR_CFEN))
+		return TEE_SUCCESS;
+
+	if (SCID_OK(cidcfgr, scid_mask, cid_to_check))
+		return TEE_SUCCESS;
+
+	if (SEM_EN_AND_OK(cidcfgr, cid_to_check)) {
+		if (!(semcr & _SEMCR_MUTEX) ||
+		    ((semcr & scid_mask) >> SCID_SHIFT) == cid_to_check) {
+			return TEE_SUCCESS;
+		}
+	}
+
+	return TEE_ERROR_ACCESS_DENIED;
+}
+
+void stm32_rif_parse_cfg(uint32_t rif_conf,
+			 struct rif_conf_data *conf_data,
+			 unsigned int nb_cid_supp,
+			 unsigned int nb_channel)
+{
+	uint32_t cidcfdg_conf_mask = 0;
+	uint32_t channel_id = 0;
+	uint32_t msb_nb_cid_supp = sizeof(nb_cid_supp) * 8 -
+				   __builtin_clz(nb_cid_supp | 1) - 1;
+	uint32_t scid_mask = 0;
+	uint32_t semwl_mask = 0;
+	unsigned int conf_index = 0;
+
+	/* SCID bitfield highend can't be > SCID_SHIFT + MAX_CID_BITFIELD */
+	assert(msb_nb_cid_supp < MAX_CID_BITFIELD);
+
+	scid_mask = GENMASK_32(SCID_SHIFT + msb_nb_cid_supp, SCID_SHIFT);
+	semwl_mask = GENMASK_32(SEMWL_SHIFT + nb_cid_supp, SEMWL_SHIFT);
+
+	cidcfdg_conf_mask = scid_mask | semwl_mask | _CIDCFGR_CFEN |
+			    _CIDCFGR_SEMEN;
+
+	/* Shift corresponding to the desired resources */
+	channel_id = RIF_CHANNEL_ID(rif_conf);
+	if (channel_id >= nb_channel)
+		panic("Bad RIF controllers number");
+
+	/*
+	 * Make sure we set the bits in the right sec and priv conf register.
+	 * This is done to support IPS having more than 32 channels.
+	 */
```

Comment:
I don't think the 1st sentence is really useful. I would replace this with:
`/* Some peripherals have more than 32 RIF channels */`